src/lib/mina_base/parties_logic.ml

```diff
@@ -428,48 +420,75 @@ module Make (Inputs : Inputs_intf) = struct
-        Opt.or_default ~if_:Call_stack.if_ ~default:Call_stack.empty
-          (Opt.map ~f:snd res)
-      in
-      (next_stack, next_call_stack)
-    in
     (* If the current stack is complete, 'return' to the previous
        partially-completed one.
     *)
-    let current_stack, call_stack =
-      let current_is_empty = Ps.is_empty current_stack in
-      ( Ps.if_ current_is_empty ~then_:next_stack ~else_:current_stack
+    let current_forest, call_stack =
+      let next_forest, next_call_stack =
+        (* Invariant: call_stack contains only non-empty forests. *)
+        let res = Call_stack.pop call_stack in
+        let next_forest =
+          Opt.or_default ~if_:Ps.if_ ~default:Ps.empty (Opt.map ~f:fst res)
+        in
+        let next_call_stack =
+          Opt.or_default ~if_:Call_stack.if_ ~default:Call_stack.empty
+            (Opt.map ~f:snd res)
+        in
+        (next_forest, next_call_stack)
+      in
+      (* TODO: I believe current should only be empty for the first party in
+         a transaction. *)
+      let current_is_empty = Ps.is_empty current_forest in
+      ( Ps.if_ current_is_empty ~then_:next_forest ~else_:current_forest
       , Call_stack.if_ current_is_empty ~then_:next_call_stack ~else_:call_stack
       )
     in
-    let stack_or_party, next_stack = Ps.pop_exn current_stack in
-    let party, remaining_stack =
-      let as_stack = Ps.as_stack stack_or_party in
-      let stack = Opt.or_default ~if_:Ps.if_ ~default:current_stack as_stack in
-      let popped_value, remaining_stack = Ps.pop_party_exn stack in
-      ( popped_value
-      , Ps.if_ (Opt.is_some as_stack) ~then_:remaining_stack ~else_:Ps.empty )
+    let (party, party_forest), remainder_of_current_forest =
+      Ps.pop_exn current_forest
+    in
+    (* Cases:
+       - [party_forest] is empty, [remainder_of_current_forest] is empty.
+       Pop from the call stack to get another forest, which is guaranteed to be non-empty.
+       The result of popping becomes the "current forest".
+       - [party_forest] is empty, [remainder_of_current_forest] is non-empty.
+       Push nothing to the stack. [remainder_of_current_forest] becomes new "current forest"
+       - [party_forest] is non-empty, [remainder_of_current_forest] is empty.
+       Push nothing to the stack. [party_forest] becomes new "current forest"
+       - [party_forest] is non-empty, [remainder_of_current_forest] is non-empty:
+       Push [remainder_of_current_forest] to the stack. [party_forest] becomes new "current forest".
+    *)
+    let party_forest_empty = Ps.is_empty party_forest in
+    let remainder_of_current_forest_empty =
+      Ps.is_empty remainder_of_current_forest
     in
-    let current_stack, next_stack =
-      let is_empty = Ps.is_empty remaining_stack in
-      ( Ps.if_ is_empty ~then_:next_stack ~else_:remaining_stack
-      , Ps.if_ is_empty ~then_:Ps.empty ~else_:next_stack )
+    let popped_call_stack = Call_stack.pop call_stack in
```

Comment:
This would benefit from similar treatment to above:
```ocaml
let (popped_forest, popped_call_stack) =
    let res = Call_stack.pop call_stack in
    (Opt.map ~f:fst res, Opt.map ~f:snd res)
in
```